File: scripts/rttnews-scraper.ts
#!/usr/bin/env bun
/**
 * RTTNews FDA Calendar Web Scraper
 * 
 * Scrapes PDUFA and FDA review dates from https://www.rttnews.com/corpinfo/fdacalendar.aspx
 * Run on the 1st of each month to update data
 */

const BASE_URL = "https://www.rttnews.com/corpinfo/fdacalendar.aspx";

interface PdufaItem {
  id: string;
  company: string;
  ticker: string;
  drug: string;
  pdufaDate: string;
  description: string;
  category: string;
  status: string;
  sourceUrl: string;
  scrapedAt: string;
}

// Parse date from MM/DD/YYYY format
function parseDate(dateStr: string): string {
  const match = dateStr.match(/(\d{1,2})\/(\d{1,2})\/(\d{4})/);
  if (!match) return "";
  const [_, month, day, year] = match;
  const paddedMonth = month.padStart(2, '0');
  const paddedDay = day.padStart(2, '0');
  return `${year}-${paddedMonth}-${paddedDay}`;
}

// Extract ticker from company name (e.g., "Regenxbio Inc (RGNX)" -> "RGNX")
function extractTicker(text: string): string {
  const match = text.match(/\(([A-Z]+)\)/);
  return match ? match[1] : "";
}

// Clean company name (remove ticker)
function cleanCompanyName(companyName: string): string {
  return companyName.replace(/\s*\([A-Z]+\)\s*/g, "").trim();
}

// Infer category from drug description
function inferCategory(description: string): string {
  const desc = description.toLowerCase();
  if (desc.includes("cancer") || desc.includes("tumor") || desc.includes("oncology") || desc.includes("leukemia") || desc.includes("prostate") || desc.includes("ovarian") || desc.includes("bladder")) return "Oncology";
  if (desc.includes("diabetes")) return "Metabolic/Endocrine";
  if (desc.includes("schizophrenia") || desc.includes("bipolar") || desc.includes("depression") || desc.includes("mental")) return "Neurology/Psychiatry";
  if (desc.includes("achondroplasia") || desc.includes("metabolic") || desc.includes("syndrome") || desc.includes("mucopolysaccharidosis")) return "Genetic/Rare Disease";
  if (desc.includes("respiratory") || desc.includes("rhinosinusitis")) return "Respiratory";
  if (desc.includes("imaging") || desc.includes("pet")) return "Diagnostics";
  return "Other";
}

async function scrapeRTTNews(): Promise<PdufaItem[]> {
  console.log(`Scraping RTTNews FDA Calendar from ${BASE_URL}...`);
  
  const response = await fetch(BASE_URL);
  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`);
  }
  
  const html = await response.text();
  
  // Parse HTML to extract data
  const items: PdufaItem[] = [];
  const lines = html.split('\n');
  
  let idCounter = 1;
  let currentCompany = "";
  let currentTicker = "";
  let currentDrug = "";
  let currentDate = "";
  let currentDescription = "";
  let foundEntry = false;
  
  for (const line of lines) {
    // Look for company name with ticker link
    const match = line.match(/symbolsearch\.aspx\?symbol=([^&]+)">([^<]+)<\/a>\s*\([^)]+\)/);
    if (match) {
      currentTicker = match[1];
      currentCompany = match[2];
      console.log(`Found company: ${currentCompany} (${currentTicker})`);
    }
    
    // Look for drug name - check for BLA, NDA, sBLA, sNDA indicators
    if (currentCompany && (line.includes("(BLA") || line.includes("(NDA") || line.includes("(sBLA") || line.includes("(sNDA") || line.includes("BLA)") || line.includes("NDA)") || line.includes("sBLA)") || line.includes("sNDA)"))) {
      // Extract drug name before the parentheses
      const drugMatch = line.match(/^([^<(]+)(?:\s*\((?:BLA|NDA|sBLA|sNDA)\))?\s*(?:\([A-Z0-9\.]+\))?\s*$/);
      if (drugMatch) {
        currentDrug = drugMatch[1].trim();
        if (currentDrug.length > 0) {
          foundEntry = true;
          console.log(`Found drug: ${currentDrug}`);
        }
      }
    }
    
    // Look for date in MM/DD/YYYY format
    if (foundEntry && !currentDate) {
      const dateMatch = line.match(/(\d{1,2}\/\d{1,2}\/\d{4})/);
      if (dateMatch) {
        currentDate = dateMatch[1];
        console.log(`Found date: ${currentDate}`);
      }
    }
    
    // Look for "FDA decision" description
    if (foundEntry && currentDate && line.includes("FDA decision")) {
      currentDescription = line.trim();
      console.log(`Found description: ${currentDescription.substring(0, 50)}...`);
      
      // Parse the date
      const parsedDate = parseDate(currentDate);
      if (!parsedDate) {
        console.log(`Invalid date format: ${currentDate}`);
        continue;
      }
      
      // Only include dates from today onwards (no backdating)
      const today = new Date();
      today.setHours(0, 0, 0, 0);
      const pdufaDate = new Date(parsedDate);
      
      if (pdufaDate < today) {
        console.log(`Skipping past date: ${parsedDate} (${currentCompany})`);
      } else {
        items.push({
          id: String(idCounter++),
          company: cleanCompanyName(currentCompany),
          ticker: currentTicker.toUpperCase(),
          drug: currentDrug,
          pdufaDate: parsedDate,
          description: currentDescription,
          category: inferCategory(currentDescription),
          status: "Pending",
          sourceUrl: BASE_URL,
          scrapedAt: new Date().toISOString()
        });
        console.log(`Added: ${currentCompany} - ${currentDrug} - ${parsedDate}`);
      }
      
      // Reset for next entry
      currentTicker = "";
      currentCompany = "";
      currentDrug = "";
      currentDate = "";
      currentDescription = "";
      foundEntry = false;
    }
  }
  
  // Sort by PDUFA date
  items.sort((a, b) => a.pdufaDate.localeCompare(b.pdufaDate));
  
  console.log(`Scraped ${items.length} upcoming PDUFA dates`);
  return items;
}

// Save data to JSON file
function saveData(data: PdufaItem[], outputPath: string) {
  const output = {
    data,
    meta: {
      source: "RTTNews",
      sourceUrl: BASE_URL,
      scrapedAt: new Date().toISOString(),
      itemCount: data.length,
      dateRange: {
        earliest: data.length > 0 ? data[0].pdufaDate : null,
        latest: data.length > 0 ? data[data.length - 1].pdufaDate : null
      }
    }
  };
  
  Bun.write(outputPath, JSON.stringify(output, null, 2));
  console.log(`Data saved to ${outputPath}`);
}

async function main() {
  try {
    // Scrape data
    const data = await scrapeRTTNews();
    
    // Save to multiple locations
    const dataPath = "/home/workspace/fda-tracker-vercel/public/fda-data.json";
    
    // Ensure directory exists
    const dirPath = "/home/workspace/fda-tracker-vercel/public";
    await Bun.$`mkdir -p ${dirPath}`;
    
    saveData(data, dataPath);
    
    console.log("\n✅ Scraping complete!");
    console.log(`📊 Total items: ${data.length}`);
    console.log(`📁 Saved to: ${dataPath}`);
    
  } catch (error) {
    console.error("❌ Error scraping RTTNews:", error);
    process.exit(1);
  }
}

main();

File: scripts/rttnews-scraper.py
#!/usr/bin/env python3
"""
RTTNews FDA Calendar Web Scraper
Scrapes PDUFA and FDA review dates from https://www.rttnews.com/corpinfo/fdacalendar.aspx
Run on the 1st of each month to update data
"""

import requests
from bs4 import BeautifulSoup
from datetime import datetime, date
import json
import re

BASE_URL = "https://www.rttnews.com/corpinfo/fdacalendar.aspx"
OUTPUT_PATH = "/home/workspace/fda-tracker-vercel/public/fda-data.json"


def parse_date(date_str):
    """Parse date from MM/DD/YYYY format to YYYY-MM-DD"""
    match = re.match(r'(\d{1,2})/(\d{1,2})/(\d{4})', date_str)
    if not match:
        return ""
    month, day, year = match.groups()
    return f"{year}-{month.zfill(2)}-{day.zfill(2)}"


def extract_ticker(text):
    """Extract ticker from company name link"""
    match = re.search(r'symbol=([A-Z]+)', text)
    return match.group(1) if match else ""


def clean_company_name(company_name):
    """Remove ticker and extra spaces from company name"""
    cleaned = re.sub(r'\s*\([^)]+\)\s*', '', company_name)
    return cleaned.strip()


def infer_category(description):
    """Infer category from drug description"""
    desc = description.lower()
    if any(x in desc for x in ["cancer", "tumor", "oncology", "leukemia", "prostate", "ovarian", "bladder"]):
        return "Oncology"
    if "diabetes" in desc:
        return "Metabolic/Endocrine"
    if any(x in desc for x in ["schizophrenia", "bipolar", "depression", "mental"]):
        return "Neurology/Psychiatry"
    if any(x in desc for x in ["achondroplasia", "metabolic", "syndrome", "mucopolysaccharidosis"]):
        return "Genetic/Rare Disease"
    if any(x in desc for x in ["respiratory", "rhinosinusitis"]):
        return "Respiratory"
    if any(x in desc for x in ["imaging", "pet"]):
        return "Diagnostics"
    return "Other"


def scrape_rttnews():
    """Scrape RTTNews FDA Calendar"""
    print(f"Scraping RTTNews FDA Calendar from {BASE_URL}...")
    
    response = requests.get(BASE_URL)
    if not response.ok:
        raise Exception(f"HTTP error! status: {response.status}")
    
    soup = BeautifulSoup(response.text, 'html.parser')
    
    items = []
    id_counter = 1
    
    # Find all company divs
    company_divs = soup.find_all('div', {'data-th': 'Company Name', 'data-group': 'calData'})
    
    for company_div in company_divs:
        # Get company name and ticker
        company_link = company_div.find('a', href=re.compile(r'symbolsearch\.aspx'))
        if not company_link:
            continue
            
        company_name = company_div.get_text(separator=' ').strip()
        ticker = extract_ticker(str(company_link))
        
        if not ticker:
            print(f"Warning: No ticker found for {company_name}")
            continue
        
        print(f"Found company: {company_name} ({ticker})")
        
        # The next sibling divs contain Drug, Event, Outcome
        # Find the next div after company_div
        next_div = company_div.find_next_sibling('div', {'data-group': 'calData'})
        
        if not next_div or next_div.get('data-th') != 'Drug':
            continue
            
        drug_div = next_div
        drug_name = drug_div.get_text().strip()
        # Remove parentheses content like (BLA), (NDA)
        drug_name = re.sub(r'\s*\([^)]+\)\s*', '', drug_name).strip()
        
        print(f"  Found drug: {drug_name}")
        
        # Next is Event div with date and description
        event_div = drug_div.find_next_sibling('div', {'data-group': 'calData'})
        
        if not event_div or event_div.get('data-th') != 'Event':
            continue
            
        # Extract date from span with bg-purple class
        date_span = event_div.find('span', class_='bg-purple')
        if not date_span:
            continue
            
        date_str = date_span.get_text().strip()
        parsed_date = parse_date(date_str)
        
        if not parsed_date:
            print(f"  Warning: Invalid date format: {date_str}")
            continue
        
        # Extract description (after the span and <br />)
        event_text = event_div.get_text(separator=' ', strip=True)
        # Remove the date portion and keep the description
        description = event_text.replace(date_str, '').strip()
        # Clean up extra whitespace
        description = re.sub(r'\s+', ' ', description).strip()
        
        print(f"  Found date: {parsed_date}")
        print(f"  Found description: {description[:50]}...")
        
        # Check if date is in the future (no backdating)
        today = date.today()
        pdufa_date = datetime.strptime(parsed_date, "%Y-%m-%d").date()
        
        if pdufa_date < today:
            print(f"  Skipping past date: {parsed_date}")
            continue
        
        # Create item
        item = {
            "id": str(id_counter),
            "company": clean_company_name(company_name),
            "ticker": ticker.upper(),
            "drug": drug_name,
            "pdufaDate": parsed_date,
            "description": description,
            "category": infer_category(description),
            "status": "Pending",
            "sourceUrl": BASE_URL,
            "scrapedAt": datetime.now().isoformat()
        }
        
        items.append(item)
        print(f"  Added: {item['company']} - {item['drug']} - {parsed_date}")
        
        id_counter += 1
    
    # Sort by PDUFA date
    items.sort(key=lambda x: x['pdufaDate'])
    
    print(f"Scraped {len(items)} upcoming PDUFA dates")
    return items


def save_data(items, output_path):
    """Save data to JSON file"""
    output = {
        "data": items,
        "meta": {
            "source": "RTTNews",
            "sourceUrl": BASE_URL,
            "scrapedAt": datetime.now().isoformat(),
            "itemCount": len(items),
            "dateRange": {
                "earliest": items[0]['pdufaDate'] if items else None,
                "latest": items[-1]['pdufaDate'] if items else None
            }
        }
    }
    
    with open(output_path, 'w') as f:
        json.dump(output, f, indent=2)
    
    print(f"Data saved to {output_path}")


def main():
    try:
        # Scrape data
        items = scrape_rttnews()
        
        # Create output directory if needed
        import os
        output_dir = os.path.dirname(OUTPUT_PATH)
        os.makedirs(output_dir, exist_ok=True)
        
        # Save data
        save_data(items, OUTPUT_PATH)
        
        print("\n✅ Scraping complete!")
        print(f"📊 Total items: {len(items)}")
        print(f"📁 Saved to: {OUTPUT_PATH}")
        
    except Exception as error:
        print(f"❌ Error scraping RTTNews: {error}")
        import traceback
        traceback.print_exc()
        exit(1)


if __name__ == "__main__":
    main()
